File: problems/222/solution.go
package main
import "math"

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func countNodes(root *TreeNode) int {
    if root == nil {
        return 0
    }
    
    cur, hLeft, hRight := root, 1, 1
    
    for cur.Left != nil {
        hLeft++
        cur = cur.Left
    }
    
    cur = root
    
    for cur.Right != nil {
        hRight++
        cur = cur.Right
    }
    
    if hLeft == hRight {
        return int(math.Pow(2.0, float64(hLeft))) - 1
    }
    
    return 1 + countNodes(root.Left) + countNodes(root.Right)
}

File: problems/222/solution.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def countNodes(self, root: TreeNode) -> int:
        if not root:
            return 0
        
        cur, lh, rh = root, 0, 0
        
        while cur:
            lh += 1
            cur = cur.left
        
        cur = root
        
        while cur:
            rh += 1
            cur = cur.right
        
        
        if lh == rh:
            return 2**lh - 1
        
        return 1 + self.countNodes(root.left) + self.countNodes(root.right)